libcnb/src/layer/public_interface.rs

```diff
@@ -0,0 +1,187 @@
+    /// The buildpack this layer is used with.
+    type Buildpack: Buildpack;
+
+    /// The metadata type for this layer. This is the data within `[metadata]` of the layer content
+    /// metadata. If you wish to use raw, untyped, TOML data instead, use [`GenericMetadata`](crate::generic::GenericMetadata).
+    ///
+    /// If the layer metadata cannot be parsed into this type, libcnb will call [`migrate_incompatible_metadata`](Self::migrate_incompatible_metadata)
+    /// with the layer's metadata as raw TOML. This allows migration of older metadata.
+    type Metadata: DeserializeOwned + Serialize + Clone;
+
+    /// Returns the types of this layer. Will be called by libcnb whenever it needs to determine
+    /// the types for this layer. This includes, but is not limited to, after create, update and
+    /// when the layer is not modified at all.
+    ///
+    /// # Implementation Requirements
+    /// Implementations **MUST** be pure. This includes that they **MUST NOT** side-effect,
+    /// including writing to stdout/stderr or the file system.
+    fn types(&self) -> LayerTypes;
+
+    /// Creates the layer from scratch.
+    ///
+    /// `layer_path` will be an empty directory where this method can write files to. Layer
+    /// metadata, including environment variables, is part of the return value of this function and
+    /// will be written to the appropriate locations by libcnb automatically.
+    ///
+    /// # Implementation Requirements
+    /// Implementations **MUST NOT** write to any other location than `layer_path`.
+    fn create(
+        &self,
+        context: &BuildContext<Self::Buildpack>,
+        layer_path: &Path,
+    ) -> Result<LayerResult<Self::Metadata>, <Self::Buildpack as Buildpack>::Error>;
+
+    /// This method will be called by libcnb when the layer already exists to determine the strategy
+    /// used to deal with it. Implementations of this method can use the current layer contents and
+    /// metadata to make that decision.
+    ///
+    /// This can be used to invalidate a layer based on metadata. For example, the layer metadata
+    /// could contain a language runtime version string. If the version requested by the user is
+    /// different, this method should can return [`ExistingLayerStrategy::Recreate`], causing a new
+    /// language runtime version to be installed from scratch.
+    ///
+    /// It can also be used cause a call to [`update`](Layer::update), updating the contents of the
+    /// existing layer. Installing additional application dependencies with a package manager is
+    /// one common case where this strategy makes sense. Implementations need to return
+    /// [`ExistingLayerStrategy::Update`] to trigger that behavior.
+    ///
+    /// When not implemented, the default implementation will return
+    /// [`ExistingLayerStrategy::Keep`], causing the layer to stay as-is (i.e. no calls to either
+    /// [`create`](Layer::create) nor [`update`](Layer::update)).
```

Comment:
So I feel like the choice of `existing_layer_strategy` default implementation return value should be based on a combination of what:
1. is safest / reduced chance of footguns
2. will mean the least boilerplate
3. is most intuitive

For classic buildpacks, we know that a massive footgun is people not invalidating the cache in cases where they should (stack change, version changes, ...), which would make the answer to (1) be `ExistingLayerStrategy::Recreate`.

In terms of reducing boilerplate, I'm not sure I can think of any scenario where a buildpack would want to unconditionally return `ExistingLayerStrategy::Keep`. That is, even if a buildpack wants to use `Keep`, it would always be in combination with another strategy, as part of a conditional in `existing_layer_strategy()`. As such, they would still need to implement `existing_layer_strategy()` themselves, therefore we haven't reduced any boilerplate.

I also don't think `ExistingLayerStrategy::Update` would be a good strategy, since not all buildpacks will need to implement `update()`, and this would also suffer from the same caching footgun as `Keep`.

If anything, we want to have no default implementation of `existing_layer_strategy()` iff `cache==true`.

One option would be to change the default `existing_layer_strategy` implementation to be just `unimplemented!("...")` - which would force people using `cache=true` to implement, without affecting the `cache=false` cases.

However one downside to that is the "the layer may still exist from before even if cache=false now" case.

Which brings us full circle to perhaps `ExistingLayerStrategy::Recreate` being the safest option for the default implementation?

Reply:
> One option would be to change the default existing_layer_strategy implementation to be just unimplemented!("...") - which would force people using cache=true to implement, without affecting the cache=false cases.

This would also have the impact of writing what looks like a functioning buildpack and having it work on the first deploy, but suddenly failing on the second. I think that's not great. I would like to avoid the "looks like it works, but it's totally broken" cases as much as possible.

Is there a way to differentiate between a default implemented trait method and a user implemented one? If so, we could check if `cache = true` and raise in `handle_layer` if the method is unimplemented. Actually, as I'm talking this out, we could just call the method and it would panic then we would know.

> Which brings us full circle to perhaps ExistingLayerStrategy::Recreate being the safest option for the default implementation?

That seems fine. By default most code runs every time, if you want to opt into caching you need to add logic. I think though a warning or error where someone is using `cache=true` with default recreate behavior would be ideal. But detecting that case is still an open question in my mind.

Reply:
I agree with all of the above, `ExistingLayerStrategy::Recreate` is the most sane default. WRT weird combinations of layer types and unimplemented methods, I think the idea of having more specific `Layer` traits such as `CachedLayer` would solve that issue. We can add those on top of the changes in this PR easily.

Reply:
Implemented in 5c99d2db16aff0a3dd636c3f7f421ec51bac6f2f